app/models/user.rb

```diff
@@ -54,22 +54,53 @@ def administrated_projects
     scope
   end
 
-  def self.to_csv
-    @users = User.order(:id)
+  def self.to_csv(
+    inherited: false, deleted: false, project_id: nil, user_id: nil,
+    datetime: (Time.now.strftime "%Y%m%d_%H%M")
+  )
+    inherited = true if project_id || user_id
+    users = if user_id.nil?
+      deleted ? User.unscoped.order(:id) : User.order(:id)
+    else
+      user_id && User.unscoped.where(id: user_id)
+    end
+    if inherited
+      permissions_projects = project_id ? Project.where(id: project_id) : Project.all
+    end
+    total = permissions_count(project_id, inherited, users, permissions_projects)
+    summary = ["-", "Generated At", datetime, "Users", users.count.to_s, "Total entries", total.to_s]
+    options_applied = [
+      "-", "Options",
+      {
+        inherited: inherited,
+        deleted: deleted,
+        project_id: project_id,
+        user_id: user_id
+      }.to_json
+    ]
+
     CSV.generate do |csv|
-      csv << [
-        "id", "name", "email", "projectiD", "project", "role", User.count.to_s + " Users",
-        (User.count + UserProjectRole.joins(:user, :project).count).to_s + " Total entries"
-      ]
-      @users.find_each do |user|
-        csv << [user.id, user.name, user.email, "", "SYSTEM", user.role.name]
-        UserProjectRole.where(user_id: user.id).joins(:project).find_each do |user_project_role|
-          csv << [
-            user.id, user.name, user.email, user_project_role.project_id,
-            user_project_role.project.name, user_project_role.role.name
-          ]
+      csv << ["id", "name", "email", "projectiD", "project", "role", "deleted at"]
+      users.find_each do |user|
```

Comment:
find_each saves a few percent of the total queries, but I guess it doesn't matter too much.